design/design-generic-references.md

```diff
@@ -0,0 +1,314 @@
+# Generic Resource References
+
+* Owner: Predrag Knezevic (@pedjak)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Background
+
+Crossplane providers provision and monitor external resources in third party
+cloud providers, following the declared specification in the corresponding managed
+resources. Very often, creation of these objects cannot happen in parallel, i.e. some
+fields can be set only after a given dependency is created.
+
+[Cross resource referencing] solves this problem for majority of cases by
+following [Kuberentes API object referemce] convention:
+
+```yaml
+spec:
+  fooRef:
+    name: bar
+```
+
+Field `spec.fooRef` contains the name of resource where the value for `spec.foo` is to
+be found. The kind of referred resource and the field containing the value used later for
+setting `spec.foo` is determined during compile time.
+
+Although the used convention covers the majority of use cases, it lacks the
+following properties:
+
+* The referred object kind and the source field cannot be tuned/changed at 
+  runtime for a given object
+* The referencer field may not always target a single referred kind. For
+  example, a URL field in Route53 could get populated by a URL of an S3 bucket
+  or access endpoint of an RDS Instance
+* The referencer field may not always target a single field in the referred
+  object
+* The value that is about to be assigned to the field require some sort of
+  transformation
+
+Some of the above issues can be solved by defining a composition. However, there
+are cases where one would like to use and refer a resource not bound by a 
+composite instance.
+
+__**NOTE**__: there is 
+[a previous work on this topic](https://github.com/crossplane/crossplane/pull/2385).
+
+## Goals
+
+The generic cross-reference should be able to:
+
+* be configurable at the runtime
+* use the existing [Kuberentes API object referemce] convention 
+  as much as possible
+* provide alternative convention in cases where [Kuberentes API object referemce]
+  convention is not suitable
+* be implemented without or light changes in RBAC
+* support referring multiple objects (namespaced and cluster-scoped) and their
+  fields
+* support value transformation before assignment
+
+## Proposal
+
+We would like to introduce a new cluster-scoped `Referable` (alternative name
```

Comment:
Could you maybe touch on why this should be cluster-scope and not namespace-scoped?

Reply:
simply because all our types are cluster-scoped, iirc.